scipy/optimize/_direct.py

```diff
@@ -0,0 +1,182 @@
+r"""
+A python wrapper to the DIRECT algorithm implemented in Fortran
+===============================================================
+
+DIRECT is a method to solve global bound constraint optimization problems and
+was originally developed by D. R. Jones, C. D. Perttunen and B. E. Stuckmann.
+It is designed to find **global** solutions of mathematical optimization 
+problems of the from,
+
+.. math::
+
+       \min_ {x \in R^n} f(x)
+
+subject to
+
+.. math::
+
+       x_L \leq  x  \leq x_U
+
+Where :math:`x` are the optimization variables (with upper and lower
+bounds), :math:`f(x)` is the objective function.
+
+The DIRECT package uses the Fortran implementation of DIRECT written by
+Joerg.M.Gablonsky, DIRECT Version 2.0.4. More information on the DIRECT
+algorithm can be found in Gablonsky's `thesis 
+<http://repository.lib.ncsu.edu/ir/bitstream/1840.16/3920/1/etd.pdf>`_.
+
+Authors:
+
+Andreas Mayer <andimscience@gmail.com>, Amit Aides <amitibo@tx.technion.ac.il>
+"""
+
+import numpy as np
+from ._directmodule import direct
+from .optimize import OptimizeResult
+
+ERROR_MESSAGES = (
+    "u[i] < l[i] for some i",
+    "maxfun is too large",
+    "Initialization failed",
+    "There was an error in the creation of the sample points",
+    "An error occured while the function was sampled",
+    "Maximum number of levels has been reached.",
+)
+
+SUCCESS_MESSAGES = (
+    "Number of function evaluations done is larger then maxfun={}",
+    "Number of iterations is equal to maxiter={}",
+    ("The best function value found is within {} percent "
+    "of the (known) global optimum"),
+    ("The volume of the hyper-rectangle with best function value found "
+     "is below volper={}"),
+    ("The volume of the hyper-rectangle with best function value found "
+     "is smaller then volper={}"),
+)
+
+def _minimize_direct(func, bounds=None, nvar=None, *args, disp=False,
+                     eps=1e-4, maxfun=20000, maxiter=6000, 
+                     locally_biased=False, fglobal=-1e100, fglper=0.01, 
+                     volper=-1.0, sigmaper=-1.0):
+    r"""
+
+    Solve an optimization problem using the DIRECT 
+    (Dividing Rectangles) algorithm.
+    It can be used to solve general nonlinear programming problems of the form:
+
+    .. math::
+
+           \min_ {x \in R^n} f(x)
+
+    subject to
+
+    .. math::
+
+           x_L \leq  x  \leq x_U
+
+    Where :math:`x` are the optimization variables (with upper and lower
+    bounds), :math:`f(x)` is the objective function.
+
+    Parameters
+    ----------
+    func : objective function
+        called as `func(x, *args)`; does not need to be defined everywhere,
+        raise an Exception where function is not defined
+    bounds : array_like, optional
+            ``(min, max)`` pairs for each element in ``x``, defining
+            the bounds on that parameter.
+    nvar : int, optional
+        Dimensionality of x (only needed if `bounds` is not defined)
+    eps : float, optional
```

Comment:
Perhaps this should be `iatol`, interval absolute tolerance.

Reply:
@dschmitz89 reminder to address this comment.

Reply:
`eps` is not used as a convergence criterion here, so I would recommend to keep the name `eps`. It seems to be used in all literature about DIRECT as well.

Reply:
Just a note about the argument that `eps` is used in the literature. Maybe, but we are making an API which should fit with our existing methods. Otherwise we would end up with different notations/variables for every methods.